blob/blob.go

```diff
@@ -165,39 +179,88 @@ func (b *Blob) UnmarshalJSON(data []byte) error {
 	b.Blob.ShareVersion = blob.ShareVersion
 	b.Commitment = blob.Commitment
 	b.namespace = blob.Namespace
+	b.index = blob.Index
 	return nil
 }
 
-// buildBlobsIfExist takes shares and tries building the Blobs from them.
-// It will build blobs either until appShares will be empty or the first incomplete blob will
-// appear, so in this specific case it will return all built blobs + remaining shares.
-func buildBlobsIfExist(appShares []shares.Share) ([]*Blob, []shares.Share, error) {
-	if len(appShares) == 0 {
-		return nil, nil, errors.New("empty shares received")
+// index represents index of the blob's first share in the eds.
+// -1 by default and only retrieved onchain blobs will have it set.
+type transformer struct {
+	index  int
+	length int
+	shares []shares.Share
+}
+
+func newTransformer(index, length int) *transformer {
+	return &transformer{
+		index:  index,
+		length: length,
 	}
-	blobs := make([]*Blob, 0, len(appShares))
-	for {
-		length, err := appShares[0].SequenceLen()
-		if err != nil {
-			return nil, nil, err
-		}
+}
 
-		amount := shares.SparseSharesNeeded(length)
-		if amount > len(appShares) {
-			return blobs, appShares, nil
+// setShares sets shares until the blob is completed and returns extra shares back.
+func (b *transformer) setShares(shares []shares.Share) (shrs []shares.Share, isComplete bool) {
```

Comment:
nit: set implies overwriting of the data everytime you call set, not adding more, so I suggest renaming to to addShares or putShares